keras_nlp/layers/cached_multi_head_attention.py

```diff
@@ -0,0 +1,116 @@
+            otherwise `(B, 1, dim)`.
+        value: Value `Tensor` of shape `(B, S, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`.
+        key: Optional key `Tensor` of shape `(B, S, dim)` if `cache=None`,
+            otherwise `(B, 1, dim)`. If not given, will use `value` for both
+            `key` and `value`, which is the most common case.
+        attention_mask: a boolean mask of shape `(B, T, S)` if `cache=None`,
+            otherwise `(B, 1, S)`. `attention_mask` prevents
+            attention to certain positions. The boolean mask specifies which
+            query elements can attend to which key elements, 1 indicates
+            attention and 0 indicates no attention. Broadcasting can happen for
+            the missing batch dimensions and the head dimension.
+        cache: a dense float Tensor. The cache of key/value of leading tokens.
+            `cache` is of shape [2, B, max_seq_len, num_heads, key_dims].
+        cache_index: a int or int Tensor, the index of the current token being
+            processed. If `cache_index=None` while `cache` is set, it means
+            it's the first pass to build the cache.
+
+    Returns:
+        attention_output: The result of the computation, of shape `(B, T, E)`,
+            where `T` is for target sequence shapes and `E` is the query input
+            last dimension if `output_shape` is `None`. Otherwise, the
+            multi-head outputs are projected to the shape specified by
+            `output_shape`.
+        cache: the updated cache.
+    """
+
+    def call(
+        self,
+        query,
+        value,
+        key=None,
+        attention_mask=None,
+        use_causal_mask=True,
+        cache=None,
+        cache_index=None,
+    ):
+        if not self._built_from_signature:
+            self._build_from_signature(query=query, value=value, key=key)
+        if key is None:
+            key = value
+
+        attention_mask = self._compute_attention_mask(
+            query,
+            value,
+            key=key,
+            attention_mask=attention_mask,
+            use_causal_mask=use_causal_mask,
+        )
+
+        query = self._query_dense(query)
+        key = self._key_dense(key)
+        value = self._value_dense(value)
+
+        if cache is not None:
+            if cache_index is None:
+                seq_len = tf.shape(query)[1]
+                k_update_indices = [0, 0, 0, 0, 0]
+                v_update_indices = [1, 0, 0, 0, 0]
+                cache_index = seq_len - 1
+            else:
+                k_update_indices = [0, 0, cache_index, 0, 0]
+                v_update_indices = [1, 0, cache_index, 0, 0]
+            cache = dynamic_update_slice(
+                cache, key[tf.newaxis, ...], k_update_indices
+            )
+            cache = dynamic_update_slice(
+                cache, value[tf.newaxis, ...], v_update_indices
+            )
+            key = cache[0, :, : cache_index + 1, :, :]
```

Comment:
Just a comment... It is still unclear to my why we need to slice here at all. Why not leave full length and rely on mask? I want to dig into this more.

Reply:
Will do this on a follow up